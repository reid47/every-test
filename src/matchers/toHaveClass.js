export default function toHaveClass(received, ...classes) {
  const isTruthy = !!received;
  const missingClasses = {};

  if (isTruthy) {
    for (let i = 0, length = classes.length; i < length; i++) {
      const className = classes[i];
      if (!received.classList.contains(className)) {
        missingClasses[className] = true;
      }
    }
  }

  return {
    pass: isTruthy && Object.keys(missingClasses).length === 0,
    isTruthy,
    missingClasses: Object.keys(missingClasses)
  };
}
