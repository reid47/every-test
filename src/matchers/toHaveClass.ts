const formatClasses = classes => classes.map(JSON.stringify).join(', ');

export function toHaveClass(received, ...classes) {
  const isTruthy = !!received;
  const missingClasses = {};

  if (isTruthy) {
    for (let i = 0, length = classes.length; i < length; i++) {
      const className = classes[i];
      if (!received.classList.contains(className)) {
        missingClasses[className] = true;
      }
    }
  }

  const missingClassNames = Object.keys(missingClasses);
  const pass = isTruthy && missingClassNames.length === 0;

  const classOrClasses = classes => (classes.length > 1 ? 'classes' : 'class');
  const message = pass
    ? `Expected element not to have ${classOrClasses(classes)} ${formatClasses(
        classes
      )}, but it did.`
    : `Expected element to have ${classOrClasses(classes)} ${formatClasses(
        classes
      )}, but it was missing ${classOrClasses(missingClassNames)} ${formatClasses(
        missingClassNames
      )}.`;

  return { pass, message };
}
