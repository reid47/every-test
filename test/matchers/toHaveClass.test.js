import toHaveClass from '../../src/matchers/toHaveClass';

describe('toHaveClass', () => {
  test('passing (single class)', () => {
    const el = document.createElement('div');
    el.classList.add('hello');

    const result = toHaveClass(el, 'hello');
    expect(result.pass).toBeTruthy();
    expect(result.message).toMatchSnapshot();
  });

  test('passing (multiple classes)', () => {
    const el = document.createElement('div');
    el.classList.add('world');
    el.classList.add('hello');

    const result = toHaveClass(el, 'hello', 'world');
    expect(result.pass).toBeTruthy();
    expect(result.message).toMatchSnapshot();
  });

  test('failing (no classes, one expected)', () => {
    const el = document.createElement('div');

    const result = toHaveClass(el, 'hello');
    expect(result.pass).toBeFalsy();
    expect(result.message).toMatchSnapshot();
  });

  test('failing (no classes, multiple expected)', () => {
    const el = document.createElement('div');

    const result = toHaveClass(el, 'hello', 'world');
    expect(result.pass).toBeFalsy();
    expect(result.message).toMatchSnapshot();
  });

  test('failing (some classes, some of expected)', () => {
    const el = document.createElement('div');
    el.classList.add('hello');

    const result = toHaveClass(el, 'hello', 'world');
    expect(result.pass).toBeFalsy();
    expect(result.message).toMatchSnapshot();
  });

  test('failing (some classes, none of expected)', () => {
    const el = document.createElement('div');
    el.classList.add('something');
    el.classList.add('else');

    const result = toHaveClass(el, 'hello', 'world');
    expect(result.pass).toBeFalsy();
    expect(result.message).toMatchSnapshot();
  });
});
