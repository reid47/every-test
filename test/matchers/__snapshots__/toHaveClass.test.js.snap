// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveClass failing (no classes, multiple expected) 1`] = `"Expected element to have classes \\"hello\\", \\"world\\", but it was missing classes \\"hello\\", \\"world\\"."`;

exports[`toHaveClass failing (no classes, one expected) 1`] = `"Expected element to have class \\"hello\\", but it was missing class \\"hello\\"."`;

exports[`toHaveClass failing (some classes, none of expected) 1`] = `"Expected element to have classes \\"hello\\", \\"world\\", but it was missing classes \\"hello\\", \\"world\\"."`;

exports[`toHaveClass failing (some classes, some of expected) 1`] = `"Expected element to have classes \\"hello\\", \\"world\\", but it was missing class \\"world\\"."`;

exports[`toHaveClass passing (multiple classes) 1`] = `"Expected element not to have classes \\"hello\\", \\"world\\", but it did."`;

exports[`toHaveClass passing (single class) 1`] = `"Expected element not to have class \\"hello\\", but it did."`;
